component/file_cache/file_cache.go

```diff
@@ -1317,6 +1317,29 @@ func (fc *FileCache) RenameFile(options internal.RenameFileOptions) error {
 	dflock.Lock()
 	defer dflock.Unlock()
 
+	// check if the source file has a dirty handle
+	var dirtySrcHandle *handlemap.Handle
+	handlemap.GetHandles().Range(
+		func(key any, value any) bool {
```

Comment:
I have two problems with such iterations:
- This will result in a crash because we are iterating it here and there can be other flows which are adding/removing items from the same map
- We do have customer cases where thousands of files are worked upon in parallel and for every rename if we go iterating the thread, it will consume too much of CPU
- Also, we are not supposed to flush the contents to storage container untill user flushes the contents manually or closes the file handle. As a result of this loop we will end up syncing data to container even for files which user has not closed yet.

Overall, this is not a safe handling.

Reply:
Thanks for taking a look at this so quickly!
I have been thinking about your concerns... Although I think [Range](https://pkg.go.dev/sync#Map.Range) is safe from crashes, I agree that my solution here is expensive and flushes a file that should not be flushed.
With those problems in mind, I came up with a new approach that should avoid both problems you pointed out. But this approach may raise new concerns, so please let me know what you think:

Instead of flushing the source file, we could call `CopyFromFile`, with `CopyFromFileOptions.Name` set to `RenameFileOptions.Dst` and `CopyFromFileOptions.File` set to the source file.
This would prevent a flush to the source file, which was not requested, but you could say it would instead "flush" the local source file to the destination file in cloud storage.

To avoid the concern you raised about the CPU cost of searching the `handlemap`, we would need to choose an approach that works without knowing whether the source file has a dirty open handle or not. To do this, I think we could check whether the source file is in the cache, using `isDownloadRequired` (is this a correct use of this function?).
- If `isDownloadRequired` returns true and the source file is not local, I think we can assume it does not have a dirty open handle, and the existing in-cloud rename process could be used in that case.
- If `isDownloadRequired` returns false and the source file is local, we could use the `CopyFromFile` process I outlined above. 

So, in a nutshell, my alternate proposal would be something like this:
```
localSrcPath := filepath.Join(fc.tmpPath, options.Src)
srcNotLocal, _, _, err := fc.isDownloadRequired(localSrcPath, options.Src, sflock)
if err != nil { ... }
if srcNotLocal {
	// rename file in cloud (previously the only code path)
	err := fc.NextComponent().RenameFile(options)
	...
} else {
	fc.policy.CacheValid(localSrcPath)
	uploadHandle, err := os.Open(localSrcPath)
	if err != nil { ... }
	err = fc.NextComponent().CopyFromFile(
		internal.CopyFromFileOptions{
			Name: options.Dst,
			File: uploadHandle,
		})
	uploadHandle.Close()
	if err != nil { ... }
}

// existing code resumes here untouched:
localDstPath := filepath.Join(fc.tmpPath, options.Dst)
...
```
What do you think of this alternate approach?

Reply:
I see what you are proposing here. In case of Rename if file does not exists locally then directly call rename on storage container. If the file exists locally then instead of doing rename on container and then locally, we just reupload the source file to container. It still has this challange that not all of local data might have flushed to local file and when we upload it migt still be stale contents as we will flush the local file when flush/close call comes. To avoid all this complexity only code was written to locally and remotely just rename the files and avoid any inconsistent uploads. Also, when we do rename on container we save a lot of cost to customer as data is copied by server from source to destination. In this approach we will end up uploading the file twice (when we rename and later if source is closed). Generally in case of editors a temp file is created and edited and at the end when you save the file its renamed and actual source gets deleted. Most editors have a config using which you can control where to keep these temp files. By default most editor tend to keep it in same directory (even in case of vim its the same). I do understand we are trying to solve the rename problem here but I see a much larger issue. When temp files are created in a mounted path we end up creating and uploading these temp files to Azure Container while these files are either going to be renamed or deleted at the end. We can save a lot of cost by keeping these temp files to some other directory instead of mounted path and that will solve this rename issue as well.

Reply:
I am not against solving the rename issue but if editor is causing this issue by creating temp files then keeping those temp files in mounted path is not a good approach as this is not a regular directory and rather a network storage. Any temp files creation will add cost to storage and network. For any network attached storage its always better to avoid such things. You can not only solve this problem but also save some $s if temp files are redirected to some other directory.

Reply:
I agree that users should not take the mounted folder to be a normal file system folder. And I agree that users should try to keep temp files outside of mounted folders. The temp file problem is definitely a headache.

I understand your objection: my proposal from October 29th would make it so that if a file is closed (and uploaded) before a rename, then it would be uploaded twice.
So the solution needs to be more specific to this situation. We are trying to detect when a file with uncommitted local writes is being renamed. But so far, I have only detected when a **_local_ file** is being renamed, which I agree is too broad.
Digging a bit deeper, I see an easy improvement to this: we could use `sflock.Count() > 0` (or maybe `sflock.Count() > 0 && !isDownloadRequired(...)`), which would detect when an **_open_ local file** is being renamed. This still isn't perfect... We don't know if the file has uncommitted writes, and we don't know if it is even open for writing. But I think this improvement would remove most of the cases where using only `isDownloadRequired` for this check would cause the software to upload the same file twice.
When we detect that an open local file is being renamed, maybe we could further reduce the risk of a redundant upload by comparing the modify date of the local file to the container's version. This adds the cost of sending a `GetAttr` to the container, but only in this already narrow case. So if that `GetAttr` call prevents a double-upload, maybe it's worth it.

What do you think of this latest proposal?

Thank you for your generous guidance as I search for a reasonable solution!

Reply:
Get-Attr calls are short-circuted. They go and fetch the data from attr-cache component and not directly from storage unless attr-cache says cache invalid. So, relying on getAttr from storage will not be helpful in all cases. This is not a simple problem to solve by just uploading as there are lot of complex scenarios involved here in case when application has multiple handles open for writing and reading. That is one of the reason to not jump on this solution quickly as we need to think through on what all can possibly go wrong with this change.

Reply:
Will try to go through the logs again and see if we are missing something which can help us sync such locally written files rename case. For now I suggest best bet is to change the default directory for temp files. Committing something quickly here is not a solution I am looking for. Need to think more on what shall be the correct approach to this problem and that will take some time and investigation. If you feel this change is good enough for you, you can create a private branch or fork out of the this repo and work with this change locally meanwhile.

Reply:
That sounds reasonable.
I will implement this approach on my fork for now, and I will look forward to your solution.
Thank you!